app/assets/javascripts/sitewide/drag_n_drop.js.erb

```diff
@@ -0,0 +1,350 @@
+    function listItems() {
+      droppedFiles = droppedFiles.filter(Boolean);
+      $('.panel-step-attachment-new').remove();
+      _dragNdropAssetsOff();
+      for(var i = 0; i < droppedFiles.length; i++) {
+        $('#new-step-assets')
+          .append(_uploadedAseetPreview(droppedFiles[i], i))
+          .promise()
+          .done(function() {
+            _removeItemHandler(i);
+          });
+      }
+      _validateTotalSize();
+      dragNdropAssetsInit('steps');
+    }
+
+    // appent the files to the form before submit
+    function appendFilesToForm(ev) {
+      var regex = /step\[assets_attributes\]\[[0-9]*\]\[id\]/;
+      var prevEls = $('input').filter(function() {
+        return this.name.match(regex);
+      });
+      droppedFiles = droppedFiles.filter(Boolean);
+      var fd = new FormData($(ev.target).closest('form').get(0));
+      for(var i = 0; i < droppedFiles.length; i++) {
+        var index = i + prevEls.length;
+        var name = 'step[assets_attributes][' + index + '][file]';
+        fd.append(name, droppedFiles[i]);
+      }
+      droppedFiles = [];
+      filesValid = true;
+      totalSize = 0;
+      _dragNdropAssetsOff();
+      return fd;
+    }
+
+    function _validateFilesSize(file) {
+      if((file.size/1048576) > <%= Constants::FILE_MAX_SIZE_MB %> && filesValid) {
+        return "<p><%= I18n.t 'general.file.size_exceeded', file_size: Constants::FILE_MAX_SIZE_MB %></p>";
+      }
+      totalSize += parseInt(file.size/1048576);
+      return '';
+    }
+
+    function _validateTotalSize() {
+      if(totalSize > <%= Constants::FILE_MAX_SIZE_MB %>) {
+        filesValid = false;
+        $.each($('.panel-step-attachment-new'), function() {
+          $(this)
+            .find('.panel-body')
+            .append("<p class='dnd-error'><%= I18n.t('general.file.total_size', size: Constants::FILE_MAX_SIZE_MB) %></p>");
+        });
+      } else {
+        $('.dnd-error').remove();
+        filesValid = true;
+      }
+    }
+
+    function _uploadedAseetPreview(asset, i) {
+      var html = '<div class="panel panel-default panel-step-attachment-new">';
+      html += '<div class="panel-heading">';
+      html += '<span class="glyphicon glyphicon-file"></span>';
+      html += '<%= I18n.t 'assets.drag_n_drop.file_label' %>';
+      html += '<div class="pull-right">';
+      html += '<a data-item-id="' + i + '" href="#">';
+      html += '<span class="glyphicon glyphicon-remove"></span></a></div></div>';
+      html += '<div class="panel-body">';
+      html += '<label class="control-label">';
+      html += '<%= I18n.t 'assets.drag_n_drop.file_label' %>:</label> ';
+      html += truncateLongString(asset.name,
+                                 <%= Constants::FILENAME_TRUNCATION_LENGTH %>);
+      html += _validateFilesSize(asset);
+      html += '</div></div>';
+      return html;
+    }
+
+    function _removeItemHandler(id) {
```

Comment:
Like I said there are some duplications but here we have separate modules and this methods is private so for now I will leave it in like it is.